### PR TITLE
feat: LT-1307 - advice footnote

### PR DESCRIPTION
**Pull request by Mark-Spiers-kainos**

### CODE REVIEW

File: cases/libraries/advice.py [hunk 1]
```diff
@@ -92,6 +93,8 @@ def construct_coalesced_advice_values(
         user=user,
         type=advice_type,
         collated_pv_grading=pv_grading,
+        footnote=break_text.join(fields["footnote"]),
+        footnote_required=True if len(fields["footnote"]) else False,
```
Comment (andytaylorkainos) [inline cases/libraries/advice.py:97]: footnote_required=len(fields["footnote"]) > 0

File: cases/libraries/post_advice.py [hunk 2]
```diff
@@ -41,7 +43,7 @@ def check_if_team_advice_exists(case, user):
 
 
 def check_refusal_errors(advice):
-    if advice["type"].lower() == "refuse" and not advice["text"]:
+    if advice.get("type") and advice["type"].lower() == "refuse" and not advice["text"]:
```
Comment (andytaylorkainos) [inline cases/libraries/post_advice.py:46]: is there a constant for "refuse"

File: cases/libraries/post_advice.py [hunk 3]
```diff
@@ -66,7 +68,10 @@ def post_advice(request, case, level, team=False):
         if not refusal_error:
             refusal_error = check_refusal_errors(advice)
 
-    serializer = CaseAdviceSerializer(data=data, many=True)
+    # we only need to know if the user has the permission if not final advice
+    footnote_permission = request.user.has_permission(GovPermissions.MAINTAIN_FOOTNOTES) and level != AdviceLevel.FINAL
+
+    serializer = CaseAdviceSerializer(data=data, many=True, context={"footnote_permission": footnote_permission,})
```
Comment (andytaylorkainos) [inline cases/libraries/post_advice.py:74]: you can remove that trailing comma

File: cases/libraries/post_advice.py [hunk 4]
```diff
@@ -66,7 +68,10 @@ def post_advice(request, case, level, team=False):
         if not refusal_error:
             refusal_error = check_refusal_errors(advice)
 
-    serializer = CaseAdviceSerializer(data=data, many=True)
+    # we only need to know if the user has the permission if not final advice
+    footnote_permission = request.user.has_permission(GovPermissions.MAINTAIN_FOOTNOTES) and level != AdviceLevel.FINAL
```
Comment (andytaylorkainos) [inline cases/libraries/post_advice.py:72]: based on your comment should it not be:

`footnote_permission = request.user.has_permission(GovPermissions.MAINTAIN_FOOTNOTES) if level != AdviceLevel.FINAL else False`

File: applications/serializers/advice.py [hunk 5]
```diff
@@ -114,6 +123,21 @@ def __init__(self, *args, **kwargs):
...
+                    self.fields["footnote"].allow_blank = False
+                if self.initial_data[0].get("footnote_required") == "False":
+                    self.fields["footnote"].allow_null = True
+                    self.initial_data[0]["footnote"] = None
+            else:
+                self.fields["footnote"].allow_null = True
+                self.fields["footnote_required"].allow_null = True
+                for i in range(0, len(self.initial_data)):
+                    self.initial_data[i]["footnote"] = None
+                    self.initial_data[i]["footnote_required"] = None
```
Comment (andytaylorkainos) [inline applications/serializers/advice.py:139]: Can you please do all of this in `def validate(data)`:
I know its a painful change, but at least you have passing tests which will ensure the change works correctly.

File: cases/tests/tests_create_advice.py [hunk 6]
```diff
@@ -97,3 +98,86 @@ def test_cannot_create_advice_when_case_in_terminal_state(self, terminal_status)
...
+
+        response = self.client.post(self.standard_case_url, **self.gov_headers, data=[data])
+
+        response_data = response.json()["advice"][0]
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(response_data["footnote"], None)
+        self.assertEqual(response_data["footnote_required"], None)
+
+    def test_cannot_create_advice_without_footnote_and_and_having_permission(self):
```
Comment (andytaylorkainos) [inline cases/tests/tests_create_advice.py:121]: you have the word `and` twice

File: applications/serializers/advice.py (hunk 5 above)
Comment (andytaylorkainos) [reply applications/serializers/advice.py:144]: i will sadly resolve this due to the nested validation :'(

File: applications/serializers/advice.py [hunk 7]
```diff
@@ -114,6 +123,26 @@ def __init__(self, *args, **kwargs):
...
+            if self.context.get("footnote_permission"):
+                self.fields["footnote_required"].required = True
+                self.fields["footnote_required"].allow_null = False
+                # if a footnote is required, we have to validate each footnote is given
+                if self.initial_data[0].get("footnote_required") == "True":
+                    self.fields["footnote"].required = True
+                    self.fields["footnote"].allow_null = False
+                    self.fields["footnote"].allow_blank = False
+                # if a footnote is not required, we remove any footnotes that may already exist on the objects.
+                if self.initial_data[0].get("footnote_required") == "False":
```
Comment (stephanieboyd) [inline applications/serializers/advice.py:137]: Could we get a function for all this footnote validation?

File: conf/constants.py [hunk 8]
```diff
@@ -19,6 +19,7 @@ class GovPermissions(Enum):
     MANAGE_ALL_ROUTING_RULES = "Manage all routing rules"
     ACTIVATE_FLAGS = "Activate and deactivate flags"
     MANAGE_PICKLISTS = "Manage picklists"
+    MAINTAIN_FOOTNOTES = "Create footnotes"
```
Comment (stephanieboyd) [inline conf/constants.py:22]: why maintain when it's create?

**Pull request merged**